cirq/ops/linear_combinations.py

```diff
@@ -315,6 +316,71 @@ def copy(self) -> 'PauliSum':
         factory = type(self)
         return factory(self._linear_dict.copy())
 
+    def expectation(self,
```

Comment:
I think the user should be required to declare the kind of state they are passing in, instead of relying on the method to infer it from the shape. Otherwise state vector tensors and density matrix tensors will be indistinguishable.

Split the expectation method into `state_vector_expectation` and `density_matrix_expectation` or similar.